py3status/modules/imap.py

```diff
@@ -65,61 +73,163 @@ class Meta:
         }
 
     def post_config_hook(self):
+        # class variables:
+        self.mail_count = None
+        self.connection = None
+        self.mail_error = None  # cannot throw self.py3.error from thread
+        self.command_tag = 0  # IMAPcommands are tagged, so responses can be matched up to requests
+
         if self.security not in ["ssl", "starttls"]:
             raise ValueError("Unknown security protocol")
 
-    def check_mail(self):
-        mail_count = self._get_mail_count()
-
-        response = {'cached_until': self.py3.time_in(self.cache_timeout)}
+        self.idle_thread = Thread(target=self._get_mail_count, daemon=True)
+        if self.use_idle:
+            self.idle_thread.start()
 
-        if mail_count is None:
+    def check_mail(self):
+        # I -- acquire mail_count
+        if self.use_idle is not False:
+            if not self.idle_thread.is_alive():
+                sleep(5)  # rate-limit thread-restarting (when network is offline)
+                self.idle_thread = Thread(target=self._get_mail_count, daemon=True)
+                self.idle_thread.start()
+            response = {'cached_until': self.py3.CACHE_FOREVER}
+        else:
+            self._get_mail_count()
+            response = {'cached_until': self.py3.time_in(self.cache_timeout)}
+        if self.mail_error is not None:
+            self.py3.log(self.mail_error, level=self.py3.LOG_ERROR)
+            self.py3.error(self.mail_error)
+
+        # II -- format response
+        if self.mail_count is None:
             response['color'] = self.py3.COLOR_BAD,
             response['full_text'] = self.py3.safe_format(
                 self.format, {'unseen': STRING_UNAVAILABLE})
-        elif mail_count > 0:
+        elif self.mail_count > 0:
             response['color'] = self.py3.COLOR_NEW_MAIL or self.py3.COLOR_GOOD
             if self.allow_urgent:
                 response['urgent'] = True
 
-        if mail_count == 0 and self.hide_if_zero:
+        if self.mail_count == 0 and self.hide_if_zero:
             response['full_text'] = ''
         else:
-            response['full_text'] = self.py3.safe_format(self.format, {'unseen': mail_count})
+            response['full_text'] = self.py3.safe_format(self.format, {'unseen': self.mail_count})
 
         return response
 
+    def _check_if_idle(self, connection):
+        supports_idle = 'IDLE' in connection.capabilities
+
+        if self.use_idle is None:
+            self.use_idle = supports_idle
+            self.py3.log("Will use {}".format('idling' if self.use_idle else 'polling'))
```

Comment:
that was on purpose. (why print it if you set it explicitly?)

Reply:
No particular reason. I set it several days ago to explicitly make sure I'm idling... Was wondering why I didn't see this line in my logs. If there is a problem, we can acknowledge it's something to do with idle or not... and to acknowledge that `use_idle = False` works too.

Reply:
hmmm. i see. any idea on how to do that cleanly (i.e. not having it shown on every reconnect)?

Reply:
Maybe if you experiment with `first_run` and move stuffs to `post_config_hook`, then you might find a way to do it cleanly... such as using `if self.first_run:` for this too.

Reply:
like this? created a new class variable `first_run` set to true in `post_config-hook` and false when we checked idling/polling.

Reply:
Yeah, that'll work. We do this for internet modules, but to print a line? Sure. You can reuse this variable to move some things into `post_config_hook` too since now you know the answer to slow modules startups.

Reply:
>  but to print a line?

what are you referring to?

> move some things into post_config_hook

there isn't much more that makes sense to put in there

Reply:
Sorry. I'm referring to printing a log line.

>there isn't much more that makes sense to put in there

If you say so, then I'll need to review this again (slowly), but I think I see it with `self.first_run`. I think I was focusing on connection objects (see other Q down there).

Reply:
ok, no worries. keep me posted ;)

Reply:
oh, one thing before i go to bed: if we'd only print if `use_idle is None`, we could spare this whole discussion ;-)

Reply:
See `netdata`'s `post_config_hook`. It's not a dealbreaker. No worries. Goodnight. :-)